### ISSUE TITLE
Runnable button breaks when using ZED_SYMBOL in language extension

### ISSUE THREAD

**Issue by anderseknert** [zed-industries/zed#12475]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Hi there! 😃 

I'm working on a [language extension](https://github.com/StyraInc/zed-rego) for [Rego](https://www.openpolicyagent.org/docs/latest/policy-language/), and it's mostly getting into a publishable state. I thought I'd try the "Runnables" feature I read about in [the blog](https://zed.dev/blog/zed-decoded-tasks#runnables), as that would be a nice addition, but I'm running into issues. I've created both a `runnables.scm` file and a `tasks.json` file:

**runnables.scm**
```
(
  (rule_head (var) @run
    (#match? @run "^test_*"))
) @rego-test
```

**tasks.json**
```json
[
  {
    "label": "opa test (single test)",
    "command": "opa",
    "args": ["test", "--run", "$ZED_SYMBOL", "$ZED_WORKTREE_ROOT"],
    "tags": ["rego-test"]
  }
]
```

I can see the button showing up in the gutter:

![Screenshot 2024-05-30 at 13 14 23](https://github.com/zed-industries/zed/assets/510711/40362b6b-f536-4583-aee1-9131b7d9c36c)

However, clicking it does nothing.

I have managed to narrow it down to the presence of `$ZED_SYMBOL` in the task. If I remove that and use any of the other variables available the button will run the task... but ZED_SYMBOL is of course the name of the test, so I kinda depend on that 😅 

Not sure if related, but the only error I see in the logs is one saying:
```
2024-05-30T13:15:54+02:00 [ERROR] unexpected item event after pane was dropped
```

### Environment

Zed: v0.137.3 (Zed)
OS: macOS 14.5.0
Memory: 64 GiB
Architecture: aarch64

**Comment by osiewicz** [comment]
Hey,
This is kind of working as expected; ZED_SYMBOL is most likely not set in your context.
ZED_SYMBOL relies on symbol outline; that is, it'll essentially be equal to the last element of the outline.
E.G:
![image](https://github.com/zed-industries/zed/assets/24362066/d29f22e4-3964-443b-9ebd-2c98e2bf3b74)
With that outline, ZED_SYMBOL would be set to `handle_input`.
However, in that repo you've shared, there is no `outline.scm` for your language, which means that there'll be no outline.

Now, what you're seeing is the effect of missing variables in task context;
1. we parse the file and find that `test_me { ... }` is runnable. ok, cool
2. we look up if there's any task bound to the tag of that location. If there is, we'll display an icon in the gutter
3. Once you click the icon on the gutter, we create context for that line; this line however does not include ZED_SYMBOL (due to missing outline.scm).
4. Therefore, we have nothing to display in the dropdown.

Now, you could just add the outline and call it a day. However, I would recommend taking a different route.
Your `runnables.scm` can create custom environment variables. Named captures (other than `run`, which - as you know from the blog post - has special treatment) are automatically promoted to environmental variables in format `ZED_CUSTOM_{capture_name}`, so e.g. the value of a capture named `foo` will be stored in `ZED_CUSTOM_foo` environment variable. 
Why would you do that instead of ZED_SYMBOL route? Well, ZED_SYMBOL gives you a bit less control over what exactly is happening and it makes it possible for changes to `outline.scm` to break your runnables detection. Therefore it is better to have your own capture.

TL;DR:
change your `runnables.scm` to:
```
((
  (rule_head (var) @run @test_name
    (#match? @run "^test_*"))
) @rego-test
(#set! tag rego-test)
)
```
(that #set! bit is there as we've changed how tags are bounds to tasks, see https://github.com/zed-industries/zed/pull/12237 for more details)
And your tasks.json to:
```
[
  {
    "label": "opa test (single test)",
    "command": "opa",
    "args": ["test", "--run", "$ZED_CUSTOM_rego-test", "$ZED_WORKTREE_ROOT"],
    "tags": ["rego-test"]
  }
]
```

**Comment by anderseknert** [comment]
Wow, thanks for such a detailed and thoughtful response, @osiewicz! Much appreciated 💯 

I should probably have prefaced my issue with a disclaimer explaining how much of this is new to me, but I'm sure that fact wasn't lost to you anyway :) While I've seen outline.scm files in many of the repos I've tried to mimic while building this, I was never sure about the purpose of that file, or really the expected contents of the `languages/$language` directory, and what's supposed to go there vs the tree-sitter repo for the language. The docs on authoring extensions say nothing more than:

> The languages directory in an extension should contain one or more directories containing languages.

Which is... not a lot :) So while I'm still curious to learn more about the outline.scm file and whether I need it for other purposes, I'll go with your recommended changes and will report back tomorrow, and I'm sure we can close this then.

From what I'm reading, the "bug" here isn't that something wasn't working as intended, but perhaps how something failed silently? While I managed to come to the conclusion about ZED_SYMBOL not resolving myself, it would have been helpful to have that explained in the logs right away, and of course ideally with a note saying _why_, or a pointer to some docs for further reading. If nothing more to help you deal less with noobs like myself :)

**Comment by osiewicz** [comment]
Hey, sure, I'd love to hear back from you. I realize the docs are far from ideal, which is why I want to spend time on making tasks docs *nice* - also for extension writers. We have an upcoming docs party which is probably when it's gonna happen.
Last but not least, I hope my OG response didn't come across as cold or anything. I wrote it in a bit of a rush and it was not my intent to say "git gid" - quite the opposite in fact. I love how excited people are about tasks and I want to help you make the most out of this feature. :)

**Comment by anderseknert** [comment]
Your response came across just perfectly :) Helping people new to OPA and Rego is a big part of my daily routine, and only rarely do I have to take on the role of a newbie myself. But this is certainly a context where I find myself in that position, and it's inspiring to see another maintainer taking the time not just to point out the problem, but get into the details to help me learn more. Thanks!

**Comment by anderseknert** [comment]
I didn't manage to get it working using the recommended approach, but adding the query to `outline.scm` worked, and that's good enough for now! Looking forward to more comprehensive docs on this in the future 😃

A side note, but I noticed that the arrows in the gutter tend to jump around as I type, even if only on a single line. Looks like a bug, but I'm not sure if it's on my end or a known issue. Anyway, the original issue is resolved. Thanks for all the help!

**Comment by osiewicz** [comment]
Hmm, that's odd; it does work for me when I apply the following patch on top of your fork (branch `tasks`):
[diff.patch](https://github.com/user-attachments/files/15540039/diff.patch)
I guess that might be because I'm using nightly, though that's still gonna become a Preview/Stable eventually, so
At the very least I'd recommend using that `#set!` bit, as that'll ensure that your task won't break within a release or two (as that's now mandatory)

**Comment by anderseknert** [comment]
Thanks @osiewicz! I'll test your patch against the next release to see if it works using that, and will have our extension updated as soon as it does.
